### ISSUE TITLE
Internal Compiler Error when using extern types

### ISSUE THREAD

**Issue by lolbinarycat** [rust-lang/rust#126814]
code that caused the ICE: https://codeberg.org/binarycat/counted_str/src/commit/e04084bfcb9f33dcf5fc44ffb84ca3684e3496b6

backtrace: [rustc-ice-2024-06-21T21_39_57-71546.txt](https://github.com/user-attachments/files/15933977/rustc-ice-2024-06-21T21_39_57-71546.txt)

**Comment by saethlin** [comment]
I've transferred this issue because it's not a bug in Miri, it's a bug in the compiler overall. Miri _is_ the compiler, until main starts. But in your repro, there's nothing in main. So this is just compiler.

**Comment by saethlin** [comment]
Smaller:
```rust
#![feature(extern_types)]

extern {
    type Opaque;
}

struct ThinDst {
    x: u8,
    tail: Opaque,
}

const C1: &ThinDst = std::mem::transmute(b"d".as_ptr());
```

**Comment by RalfJung** [comment]
ICE message (which as cut off from the original backtrace -- please pots the full error message, not just a pat of it :)

```
error: internal compiler error: compiler/rustc_const_eval/src/interpret/validity.rs:1039:17: Unexpected error during validation: `extern type` does not have a known offset
```

We should really reject `extern type`s being used like that, they violate fundamental language invariants...

**Comment by RalfJung** [comment]
Note that this is not a good idea IMO:
```
// use extern types to satisfy stacked borrow rules.
#[cfg(miri)]
extern {
	type TrailingData;
}
```

Using extern types basically intimidates Miri into not checking your code, but that means Miri may miss bugs.

A better fix is to check your code with `-Zmiri-tree-borrows`, which has proper support for references that only cover the "header" of a struct. Tree Borrows should still find all UB bugs that can affect your code today, but in the future it's likely we are going to pick aliasing rules that are stricter than Tree Borrows (and weaker than Stacked Borrows).

**Comment by lolbinarycat** [comment]
> Note that this is not a good idea IMO:
> 
> ```
> // use extern types to satisfy stacked borrow rules.
> #[cfg(miri)]
> extern {
> 	type TrailingData;
> }
> ```
> 
> Using extern types basically intimidates Miri into not checking your code, but that means Miri may miss bugs.
> 
> A better fix is to check your code with `-Zmiri-tree-borrows`, which has proper support for references that only cover the "header" of a struct. Tree Borrows should still find all UB bugs that can affect your code today, but in the future it's likely we are going to pick aliasing rules that are stricter than Tree Borrows (and weaker than Stacked Borrows).

I used extern types because it's what URLO suggested to accurately represent my semantics.  it's a DST that exists behind a thin pointer.  unfortunatly the "unknown alignment" part isn't really what i wanted.  apparently tree borrows are weak enough that it allows writing to a zero-sized field past the end of a struct, so long as its backed by a larger allocation.
